apiserver/facades/agent/uniter/uniter.go

```diff
@@ -1645,7 +1681,18 @@ func (u *UniterAPI) NetworkInfo(args params.NetworkInfoParams) (params.NetworkIn
 		}
 	}
 
-	networkInfos := machine.GetNetworkInfoForSpaces(spaces)
+	// Don't get the network info for the default space if there was an error
+	// getting the address.
+	var defaultSpaceNetworkErr error
+	if spaceAddressErr != nil && spaces.Contains(environs.DefaultSpaceName) {
+		spaces.Remove(environs.DefaultSpaceName)
+		defaultSpaceNetworkErr = errors.Annotatef(err, "getting machine %q preferred primary address", machine.MachineTag())
+	}
+
+	networkInfos := machine.GetNetworkInfoForSpaces(spaces, defaultSpaceAddress)
```

Comment:
Presumably this is because we want to override the 'default space address'. I feel like this loop becomes invalid with Ian's change:
			if spaces.Contains(environs.DefaultSpaceName) && defaultSpaceAddress.Value == addr.Value() {
				r := results[environs.DefaultSpaceName]
				r.NetworkInfos, err = addAddressToResult(r.NetworkInfos, addr)
				if err != nil {
					r.Error = err
				} else {
					results[environs.DefaultSpaceName] = r
				}
			}

Because if you're using a Public address, that is likely *not* part of the machine description, and thus *no* addr.Value can == defaultSpaceAddress.Value 

It would seem that the better fix would be to just call machine.GetNetworkInfoForSpaces, and then after that call set results[DefaultSpaceName] = OverriddenDefaultAddress
or something to that effect.

I'll still consider whether we should be changing the address for the "default space" because of  a current relation. We should be changing the default value of the private address, certainly, but it doesn't feel like we're *should* be changing the address for the "default space" in what we send to the charm.
Do we know where the data is being used to just take the default value? Maybe it is unit-get private-address (which doesn't support bindings). 

Anyway, I agree with Andrew that changing Machine.GetNetworkInfoForSpaces doesn't feel like the correct fix.
We only know about CMR in the context of a relation, and its a little odd that a relation would have context on other bindings that are not its own relation. So I'm not positive where it would sit. I would tend to do something with Relation.GetNetworkInfo, which could call down to its Unit which would talk to the machine.
But it doesn't feel like we should be jumping through those layers to go directly to the machine with CMR information.